### PR TITLE
introduce 'alpha bootstrap' subcommand

### PR DESCRIPTION
**Pull request by joelanford**
This new subcommand enables two new rukpak use cases:
1. We can bootstrap rukpak as a BundleDeployment, meaning we can manage
   upgrades of rukpak itself by pivoting a 'rukpak' BundleDeployment.

   For example:
     ```
     cat rukpak-release.yaml | rukpakctl alpha bootstrap rukpak
     ```

2. We can support adoption of existing cluster objects.

   For example:
     ```
     kubectl apply -f stuff.yaml
     kubectl get -f stuff.yaml -o yaml | rukpakctl alpha bootstrap stuff
     ```

### CODE REVIEW

**Comment by tylerslaton** [comment]
@joelanford @awgreene would you say that this closes #501? I'm reviewing the code now.

**Comment by tylerslaton** [comment]
Running into a weird issue where trying to run the command you listed in the examples pends infinitely. Maybe I'm misunderstanding how this bootstrap functionality is intended to work. Does this expect that the cluster has rukpak already installed and we're just going to have it watch itself?

**Comment by joelanford** [comment]
@tylerslaton I would not expect it to pend indefinitely. Things I can think of are:
1. Something is being passed in on standard input, right?
2. I _have_ seen this thing hang due to some bugs in the upload code. I think those are ironed out with the changes in the portforward and uploader code in this PR. It probably makes sense to break those out into a separate PR, which I'll do next week (or maybe over the weekend if I'm feeling ambitious).

**Comment by tylerslaton** [comment]
Gotcha, I'll try again with that information in mind to see if I can get it working.

File: cmd/rukpakctl/cmd/alpha_bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+			}
+		},
+	}
+	cmd.Flags().StringVar(&systemNamespace, "system-namespace", "rukpak-system", "Namespace in which the core rukpak provisioners are running.")
+	cmd.Flags().StringVar(&uploadServiceName, "upload-service-name", "core", "the name of the service of the upload manager.")
+	cmd.Flags().StringVar(&caSecretName, "ca-secret-name", "rukpak-ca", "the name of the secret in the system namespace containing the root CAs used to authenticate the upload service.")
+	return cmd
+}
+
+func adoptObject(obj client.Object, systemNamespace, bundleDeploymentName string) {
```
Comment (timflannagan) [inline cmd/rukpakctl/cmd/alpha_bootstrap.go:121]: Potentially useful in the util package?

**Review by timflannagan — COMMENTED** [review]
LGTM at the surface. I think my preference here would be to move the refactoring changes + the port-forwarding fixes into their own commits or a separate PR entirely.

**Comment by joelanford** [comment]
Upload fixes were broken out into #515, and I have rebased this PR to pick those changes up. End result is that this PR is now solely focused in supporting `alpha bootstrap`

Note: I decided to leave the `type Run struct` re-factoring in this PR because this is the PR that introduces the need for that code to be shared.

Only remaining step (pending more review comments) is to move the "adoptObject" function to the util package.

**Comment by joelanford** [comment]
> Only remaining step (pending more review comments) is to move the "adoptObject" function to the util package.

Done!

File: internal/util/adopt.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package util
+
+import "sigs.k8s.io/controller-runtime/pkg/client"
+
+func AdoptObject(obj client.Object, systemNamespace, bundleDeploymentName string) {
```
Comment (timflannagan) [inline internal/util/adopt.go:5]: nit: godoc.

**Review by timflannagan — APPROVED** [review]

File: internal/util/adopt.go (hunk 2 above)
Comment (joelanford) [reply internal/util/adopt.go:5]: doh!

File: internal/util/adopt.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	obj.SetAnnotations(annotations)
+
+	labels := obj.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+
+	labels["app.kubernetes.io/managed-by"] = "Helm"
+	labels["core.rukpak.io/owner-kind"] = "BundleDeployment"
+	labels["core.rukpak.io/owner-name"] = bundleDeploymentName
```
Comment (timflannagan) [inline internal/util/adopt.go:36]: nit: use the constant variables for these labels?

**Pull request merged**